### PR TITLE
Add New simulator to dev tools & whole dev tools cleanup

### PR DESCRIPTION
**Pull request by jack60612**
- [x] Add All Commands listed below
Start ( Simulation (Start through daemon) ) 
Stop ( -r for restart) 
Farm ( x number of blocks, default 1, transaction block, default true) 
Autofarm., (On, Off) Will farm block when transaction is in mempool 
Create ( Configure Sim Chain) in chia root file, full setup wizard, Create new key prompt, start with -s ) 
Status ( keychain info, other blockchain info, key info): 
Revert ( Num blocks, default 1) 

- [x] Add Docker Image ( In chia-simulator-docker) 

- [x] #37 
- [x] https://github.com/Chia-Network/chia-blockchain/pull/12072
- [x]  https://github.com/Chia-Network/chia-blockchain/pull/12549
- [x] https://github.com/Chia-Network/chia-blockchain/pull/12521
- [x] Chia 1.5.1 Release

Tests will continue to fail until 1.5.1 is released

### CODE REVIEW

File: cdv/cmds/chia_inspect.py [hunk 1]
```diff
@@ -680,8 +676,8 @@ def one_or_zero(value):
 
             if kwargs["synthetic"]:
                 if sk:
-                    sk = calculate_synthetic_secret_key(sk, hexstr_to_bytes(kwargs["hidden_puzhash"]))
-                pk = calculate_synthetic_public_key(pk, hexstr_to_bytes(kwargs["hidden_puzhash"]))
+                    sk = calculate_synthetic_secret_key(sk, bytes32(hexstr_to_bytes(kwargs["hidden_puzhash"])))
```
Comment (altendky) [inline cdv/cmds/chia_inspect.py:679]: `bytes32.from_hexstr()`

Comment (jack60612) [reply cdv/cmds/chia_inspect.py:679]: that was the linter, but ill fix it while im here :)

Comment (jack60612) [reply cdv/cmds/chia_inspect.py:679]: oh wait im blind. thanks!

**Comment by jack60612** [comment]
all non directly sim related changes are being moved into another pr.

File: cdv/cmds/sim_utils.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    config["full_node"]["peer_connect_interval"] = 50
+    config["full_node"]["sanitize_weight_proof_only"] = False
+    config["logging"]["log_stdout"] = True
+    config["selected_network"] = "testnet0"
+    # make sure we don't try to connect to other nodes.
+    config["full_node"]["introducer_peer"] = None
+    config["wallet"]["introducer_peer"] = None
+    config["full_node"]["dns_servers"] = []
+    config["wallet"]["dns_servers"] = []
+    for service in ["harvester", "farmer", "full_node", "wallet", "introducer", "timelord", "pool", "simulator", "ui"]:
```
Comment (jack60612) [inline cdv/cmds/sim_utils.py:89]: remove simulator from here

Comment (jack60612) [reply cdv/cmds/sim_utils.py:89]: custom ports too? daemon port, rpc port, node port

**Comment by jack60612** [comment]
Sorry for the force push, easiest way to resolve conflicts

File: cdv/cmds/sim_utils.py [hunk 3]
```diff
@@ -280,10 +294,25 @@ async def async_config_wizard(
...
+    # now we create make sure the simulator has a genesis block
+    print("Please wait, generating genesis block.")
+    while True:
+        try:
+            num_blocks = await execute_with_simulator(None, root_path, get_current_height, False)
+        except ClientConnectorError:
+            await asyncio.sleep(0.25)
+        else:
+            if num_blocks == 0:
+                await execute_with_simulator(None, root_path, farm_blocks, False, 1, True, farming_address)
```
Comment (mariano54) [inline cdv/cmds/sim_utils.py:334]: Should we just return here since we are guaranteed to be height 1

Comment (jack60612) [reply cdv/cmds/sim_utils.py:334]: The create command can also be used to modify an existing simulator, so I think this should stay.

**Comment by jack60612** [comment]
Docker might be moved to a new repo depending on workflow requirements

**Comment by jack60612** [comment]
> Docker might be moved to a new repo depending on workflow requirements

docker was moved

**Comment by jack60612** [comment]
ci

File: cdv/cmds/sim_utils.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+
+
+# based on execute_with_node
+async def execute_with_simulator(
+    rpc_port: Optional[int], root_path: Path, function: Callable, catch_errors: bool = True, *args
+) -> Any:
+    import traceback
+
+    from chia.util.config import load_config
+    from chia.util.ints import uint16
```
Comment (Quexington) [inline cdv/cmds/sim_utils.py:35]: Is there a reason you're doing imports inside of the function?  These don't seem like things that matter to be optional so it seems like you're just slowing down the runtime by doing this.

File: cdv/cmds/sim_utils.py [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+def print_coin_record(
+    name: str,
+    address_prefix: str,
+    mojo_per_unit: int,
+    coin_record: CoinRecord,
+) -> None:
+    from datetime import datetime
+
+    chia_amount = Decimal(int(coin_record.coin.amount)) / mojo_per_unit
```
Comment (Quexington) [inline cdv/cmds/sim_utils.py:351]: Not a big fan of this, you're likely only looking at coin records if you're doing some wallet or chialisp development and you're likely going to need to convert to mojos anyway.  Same with the address conversion.  I have no problems with throwing them as additional information, but not replacing puzzle hash and amount.

File: cdv/cmds/sim_utils.py [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+    include_reward_coins: bool,
+    include_spent: bool = False,
+) -> None:
+    import sys
+
+    from chia.cmds.units import units
+
+    coin_records: List[CoinRecord] = await node_client.get_all_coins(include_spent)
+    coin_records = [coin_record for coin_record in coin_records if not coin_record.coinbase or include_reward_coins]
+    address_prefix = "txch"  # we are never on mainnet
```
Comment (Quexington) [inline cdv/cmds/sim_utils.py:374]: Even if we're never on mainnet, we shouldn't prevent people from using whatever prefix they want.

**Review by Quexington — APPROVED** [review]
A few small comments but otherwise looks good.

File: cdv/cmds/sim_utils.py (hunk 5 above)
Comment (jack60612) [reply cdv/cmds/sim_utils.py:351]: the entire point of this simulator is for wallet or chialisp development.

Comment (Quexington) [reply cdv/cmds/sim_utils.py:351]: That's my point, wallet/chialisp development isn't going to ever want the decmials only the mojo amounts

Comment (jack60612) [reply cdv/cmds/sim_utils.py:351]: > Not a big fan of this, you're likely only looking at coin records if you're doing some wallet or chialisp development and you're likely going to need to convert to mojos anyway. Same with the address conversion. I have no problems with throwing them as additional information, but not replacing puzzle hash and amount.
Oh i 100% over thought that, i'll fix that shortly.

File: cdv/cmds/sim_utils.py (hunk 4 above)
Comment (jack60612) [reply cdv/cmds/sim_utils.py:35]: ill move them out

File: tests/cmds/object_files/coinrecords/coinrecord.json [hunk 7]
```diff
@@ -1,7 +1,11 @@
...
+   "coin":{
+      "parent_coin_info":"0x0000000000000000000000000000000000000000000000000000000000000000",
+      "puzzle_hash":"0x0000000000000000000000000000000000000000000000000000000000000000",
+      "amount":0
+   },
+   "confirmed_block_index":1,
+   "spent_block_index":1,
+   "coinbase":true,
+   "timestamp":909469800
+}
```
Comment (Quexington) [inline tests/cmds/object_files/coinrecords/coinrecord.json:11]: Can you elaborate why you made these changes?  The point behind them being different was to test that the inspect command was working correctly with or without `0x`

**Review by Quexington — CHANGES_REQUESTED** [review]

File: tests/cmds/object_files/coinrecords/coinrecord.json (hunk 7 above)
Comment (jack60612) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: With 1.5.1 tests fail without these changes

Comment (Quexington) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: That would be indicative of a error then, and we shouldn't just abandon the flexibility right?

Comment (jack60612) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: Ah, let me try to re add the functionality

Comment (jack60612) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: i did not realize it was intended.

Comment (jack60612) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: uh @Quexington https://github.com/Chia-Network/chia_rs/blob/f5fdb37b8c62889695b76ccc9a1fb76caabf1c48/wheel/src/from_json_dict.rs#L18

Comment (Quexington) [reply tests/cmds/object_files/coinrecords/coinrecord.json:11]: Okay I'm curious why that is in there but I suppose I'll defer to the upstream

**Review by Quexington — DISMISSED** [review]

**Review by Quexington — APPROVED** [review]

**Pull request merged**